File: Modules/api-service.js
//import { getQueryParams } from "./full-story.js";
export class ApiService
{
    constructor()
    {
 //#region This will be used if the backend is not available
/*
        this.urls = [
        {
            source: 'MIA',
            feedUrl: 'https://mia.mk/feed',
            title: 'title',
            description: 'description',
            link: 'link',
            author: 'author',
            pubDate: 'pubDate',
            urlToImage: { query: 'enclosure', attribute: 'url' }
        },
        {
            source: 'Telma',
            feedUrl: 'https://telma.com.mk/feed/',
            title: 'title',
            description: 'content:encoded',
            link: 'link',
            author: 'dc:creator',
            pubDate: 'pubDate',
            urlToImage: { query: 'content\\:encoded, encoded', regex: /<img[^>]*src="([^"]*)"/i }
        },
        {
            source: '24 Vesti',
            feedUrl: 'https://admin.24.mk/api/rss.xml',
            title: '',
            description: 'content',
            link: 'link',
            author: 'author',
            pubDate: 'pubDate',
            urlToImage:  { query: 'img', attribute: 'src' }
        },
        {
            source: 'Sitel',
            feedUrl: 'https://sitel.com.mk/rss.xml',
            title: 'title',
            description: 'description',
            link: 'link',
            author: 'dc:creator',
            pubDate: 'pubDate',
            urlToImage: { query: 'description', regex: /<img[^>]*src="([^"]*)"/i }
        },
        {
            source: 'Kanal5',
            feedUrl: 'https://kanal5.com.mk/rss.aspx',
            title: 'title',
            description: 'content',
            link: 'link',
            author: '',
            pubDate: 'pubDate',
            urlToImage: { query: 'thumbnail' }
        }];
*/
//#endregion
        this.jsonUrl = 
        {
            0: 'https://localhost:5002/News',
            1: 'https://dnisko.ddns.net:5002/News'
        }
    }
    
    async fetchRssFeed()
    {
        try
        {
            let response;
            try
            {
                response = await fetch(this.jsonUrl[0]);
                console.log(response);
                if(!response.ok)
                {
                    console.warn(`Failed to fetch from ${this.jsonUrl[0]} with status ${response.status}`);
                    throw new Error(`Failed to fetch from ${this.jsonUrl[0]}`);
                }
            }
            catch(error)
            {
                console.warn(`First fetch attempt failed: ${error.message}`);
                response = await fetch(this.jsonUrl[1]);
                
                if (!response.ok)
                {
                    console.error(`Failed to fetch from ${this.jsonUrl[1]} with status ${response.status}`);
                    throw new Error(`Failed to fetch from ${this.jsonUrl[1]}`);
                }
            }
            console.log(response);
            const jsonData = await response.json();
            console.log(jsonData);
            return jsonData;

            /* //THIS WILL BE USED WHEN THE BACKEND IS NOT AVAILABLE
            // const responses = await Promise.all(this.urls.map(urlConfig => fetch(urlConfig.feedUrl)));
            // const xmlData = await Promise.all(responses.map(response => response.text()));

            // const allArticles = [];
            // for (let i = 0; i < xmlData.length; i++)
            // {
            //     const parsedArticles = this.parseRss(xmlData[i], this.urls[i]);
            //     allArticles.push(...parsedArticles);
            // }

            // console.log(allArticles);
            // return allArticles;
            */
        }
        catch (error)
        {
            console.error('Error fetching RSS feed:', error);
            return [];
        }
    }

//#region If there is a need to fetch JSON data + XML data from frontend, we use this
/*  //Fetch JSON data from url OR path to file
    async fetchJsonData()
    {
        try
        {
            const response = await fetch(this.jsonUrl);
            if (!response.ok) 
            {
                throw new Error('Failed to fetch JSON data');
            }
            const jsonData = await response.json();
            console.log('Fetched JSON data:', jsonData);
            return jsonData;
        }
        catch (error)
        {
            console.error('Error fetching JSON data:', error);
            return [];
        }
    }
*/
/*  //Fetch all news from RSS feed and JSON data
    async fetchAllNews()
    {
        try
        {
            const [rssData, jsonData] = await Promise.all([
                this.fetchRssFeed(),
                this.fetchJsonData()
            ]);

            const mergedData = this.deDuplicateNews([...rssData, ...jsonData]);
            console.log('Merged and deduplicated news data:', mergedData);
            return mergedData;
        }
        catch (error)
        {
            console.error('Error fetching all news:', error);
            return [];
        }
    }
*/
    //obsolite code once the backend is available
/*  //Distinct articles based on title, publish date, and source form the RSS feed and the JSON data
    deDuplicateNews(newsData)
    {
        
        const uniqueArticles = [];
        const seenArticles = new Set();
    
        for (const article of newsData)
        {
            const uniqueKey = `${article.title}-${article.publishDate}-${article.source}`;
            if (seenArticles.has(uniqueKey))
            {
                continue; // Skip the rest of the loop if a duplicate is found
            }
            seenArticles.add(uniqueKey);
            uniqueArticles.push(article);
        }
    
        return uniqueArticles;
    }
*/

//obsolite code once the backend is available
/*  //parses the RSS feed and returns an array of articles
    
    parseRss(xml, urlConfig)
    {
        const filteredXml = xml.replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '');
        const parser = new DOMParser();
        const xmlDoc = parser.parseFromString(filteredXml, 'text/xml');
        const items = xmlDoc.getElementsByTagName('item');
        const articles = [];

        for (let i = 0; i < items.length; i++)
        {
            const item = items[i];
            if (item.getElementsByTagName('script').length > 0)
            {
                continue;
            }
            //debugger;
            //console.log(urlConfig);
            const title = this.getElementText(item, urlConfig.title);
            let description = this.stripHtmlTags(this.getElementText(item, urlConfig.description));
            const url = this.getElementText(item, urlConfig.link);
            const author = this.getElementText(item, urlConfig.author);
            const publishDate = this.getElementText(item, urlConfig.pubDate);
            const urlToImage = this.getEnclosureUrl(item, urlConfig.urlToImage);

            articles.push(
            {
                title,
                description,
                url,
                author,
                publishDate,
                urlToImage
            });
        }
        console.log("Articles");
        console.log(articles);
        return articles;
    }
*/

//obsolite code once the backend is available
/*  //Stripping the HTML tags from the fetched RSS feed
    stripHtmlTags(text, allowedTags = [])
    {
        if (allowedTags.length === 0)
        {
            return text.replace(/<\/?[^>]+(>|$)/g, '');
        }
        
        const tagList = allowedTags.join('|');
        const regex = new RegExp(`<(?!\/?(?:${tagList})\\b)[^>]*>`, 'gi');
        return text.replace(regex, '');
    }
*/

//obsolite code once the backend is available
/*  //Get the text content of an element in the XML
    getElementText(parent, tagName)
    {
        const element = parent.getElementsByTagName(tagName)[0];
        return element ? element.textContent.trim() : '';
    }
*/

//obsolite code once the backend is available
/*  //Get the enclosure (img) URL of an item in the XML
    getEnclosureUrl(item, imageConfig)
    {
        const element = item.querySelector(imageConfig.query);
        if (element)
        {
            if (imageConfig.attribute)
            {
                return element.getAttribute(imageConfig.attribute) || '';
            }
            if (imageConfig.regex)
            {
                const match = imageConfig.regex.exec(element.textContent);
                if (match && match[1])
                {
                    return match[1];
                }
            }
            // Directly return the text content if no attribute or regex is specified
            return element.textContent.trim();
        }
        return '';
    }
*/
//#endregion

/*   //TEST CODE
    getEnclosureUrl2(item)
    {
        // Debugging and logging
        //debugger;
        console.log(item);

        // Check for enclosure tag first
        const enclosureElement = item.querySelector('enclosure');
        if (enclosureElement)
        {
            const enclosureUrl = enclosureElement.getAttribute('url');
            if (enclosureUrl)
            {
                return enclosureUrl;
            }
        }

        const imgElement = item.querySelector('img');
        if (enclosureElement)
        {
            const enclosureUrl = enclosureElement.getAttribute('src');
            if (enclosureUrl)
            {
                return enclosureUrl;
            }
        }

        // If no enclosure tag is found, check for content:encoded
        const contentEncodedElement = item.querySelector('content\\:encoded, encoded');
        if (contentEncodedElement)
        {
            const contentEncoded = contentEncodedElement.textContent;
            const imgTagRegex = /<img[^>]*src="([^"]*)"/i;
            const match = imgTagRegex.exec(contentEncoded);
            if (match && match.length > 1)
            {
                return match[1]; // This will be the URL from the src attribute of the img tag
            }
        }

        // Return an empty string if no URL is found
        return '';
        // debugger;
        // console.log(item);
        // console.log(item.lastElementChild.enclosure.url);
        // const contentEncoded = item.querySelector('content\\:encoded, encoded').textContent ?
        //                        item.querySelector('content\\:encoded, encoded').textContent : item.getElementsByTagName("enclosure").textContent;
        
        // const imgTagRegex = /<img[^>]*src="([^"]*)"/i;
        // const match = imgTagRegex.exec(contentEncoded);
        // if (match && match.length > 1)
        // {
        //     return match[1]; // This will be the URL from the src attribute of the img tag
        // }
        // return '';
    }
*/

/*   //TEST CODE
    getEnclosureUrl1(item)
    {
        const contentEncoded = item.getElementsByTagName('content:encoded')[0];
        if (contentEncoded)
        {
            const enclosureMatch = contentEncoded.textContent.match(/<enclosure\s+url="([^"]+)"/);
            if (enclosureMatch)
            {
                return enclosureMatch[1];
            }
        }
        return '';
    }
*/

/*   //TEST CODE
    filterNewsByQuery(newsData, queryParams)
    {
        if (queryParams.id)
        {
            return newsData.filter(article => article.id.toString() === queryParams.id);
        }
        return newsData;
    }
*/
}

File: Server/Services/ApiService.cs
using System.Text.RegularExpressions;
using DomainModels;
using System.Xml.Linq;

namespace Services
{
    public class ApiService
    {
        private List<UrlConfig> urls;

        public ApiService()
        {
           urls = new List<UrlConfig>
            {
                new UrlConfig
                {
                    Source = "MIA",
                    FeedUrl = "https://mia.mk/feed",
                    Title = "title",
                    Description = "description",
                    Link = "link",
                    Author = "author",
                    PubDate = "pubDate",
                    UrlToImage = new List<string> { "enclosure", "url" }
                },
                new UrlConfig
                {
                    Source = "Telma",
                    FeedUrl = "https://telma.com.mk/feed/",
                    Title = "title",
                    Description = "content:encoded",
                    Link = "link",
                    Author = "dc:creator", // Example where author tag may vary
                    PubDate = "pubDate",
                    UrlToImage = new List<string> { "content:encoded", @"<img[^>]*src=\""([^\""]*)\""" } //<img[^>]*src=\"([^\"]*)\"
                },
                new UrlConfig
                {
                    Source = "24Vesti",
                    FeedUrl = "https://admin.24.mk/api/rss.xml",
                    Title = "title",
                    Description = "content",
                    Link = "link",
                    Author = "",
                    PubDate = "pubDate",
                    UrlToImage = new List<string> { "img", "src" },
                },
                new UrlConfig
                {
                    Source = "Sitel",
                    FeedUrl = "https://sitel.com.mk/rss.xml",
                    Title = "title",
                    Description = "description",
                    Link = "link",
                    Author = "dc:creator",
                    PubDate = "pubDate",
                    UrlToImage = new List<string> { "description", @"<img[^>]*src=\""([^\""]*)\""" }
                },
                new UrlConfig
                {
                    Source = "Kanal5",
                    FeedUrl = "https://kanal5.com.mk/rss.aspx",
                    Title = "title",
                    Description = "content",
                    Link = "link",
                    Author = "author", // Example where author tag may be different
                    PubDate = "pubDate",
                    UrlToImage = new List<string> { "thumbnail", "" }
                }
            };
        }

        public async Task<List<Article>> FetchRssFeedsAsync()
        {
            var allArticles = new List<Article>();

            foreach (var urlConfig in urls)
            {
                try
                {
                    var xmlData = await FetchRssFeedXmlAsync(urlConfig.FeedUrl);
                    var parsedArticles = ParseRss(xmlData, urlConfig);
                    allArticles.AddRange(parsedArticles);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error fetching or parsing RSS feed from {urlConfig.Source}: {ex.Message}");
                }
            }

            return allArticles;
        }

        private async Task<string> FetchRssFeedXmlAsync(string feedUrl)
        {
            //using (var httpClient = new HttpClient())
            //{
            //    var response = await httpClient.GetAsync(feedUrl);
            //    response.EnsureSuccessStatusCode();
            //    return await response.Content.ReadAsStringAsync();
            //}
            using (var httpClient = new HttpClient())
            {
                // Set the User-Agent header to your desired value
                httpClient.DefaultRequestHeaders.Add("User-Agent", "RSSFetcher/1.0");

                var response = await httpClient.GetAsync(feedUrl);
                response.EnsureSuccessStatusCode();

                return await response.Content.ReadAsStringAsync();
            }
        }

        private List<Article> ParseRss(string xmlData, UrlConfig urlConfig)
        {
            var articles = new List<Article>();

            try
            {
                xmlData = RemoveScriptTags(xmlData);
                var xmlDoc = XDocument.Parse(xmlData);

                foreach (var item in xmlDoc.Descendants("item"))
                {
                    if (item.Descendants("script").Any())
                    {
                        continue;
                    }

                    var article = new Article
                    {
                        Title = GetElementValue(item, urlConfig.Title),
                        Description = StripHtmlTags(GetElementValue(item, urlConfig.Description)),
                        Link = GetElementValue(item, urlConfig.Link),
                        Author = GetElementValue(item, urlConfig.Author),
                        PubDate = GetElementValue(item, urlConfig.PubDate),
                        UrlToImage = ExtractImageUrl(item, urlConfig.UrlToImage)
                    };
                    articles.Add(article);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error parsing RSS feed: {ex.Message}");
            }

            return articles;
        }

        private string GetElementValue(XElement parent, string tagName)
        {
            if (tagName == "")
            {
                return "";
            }

            XElement element = FindElementByTagName(parent, tagName);
            return element?.Value.Trim() ?? string.Empty;
        }

        private XElement FindElementByTagName(XElement parent, string tagName)
        {
            if (tagName.Contains(":"))
            {
                var parts = tagName.Split(':');
                var prefix = parts[0];
                var localName = parts[1];
                var ns = parent.GetNamespaceOfPrefix(prefix);
                return parent.Descendants(ns + localName).FirstOrDefault();
            }
            else
            {
                return parent.Descendants(tagName).FirstOrDefault();
            }
        }
        private bool IsValidRegex(string pattern)
        {
            if (string.IsNullOrWhiteSpace(pattern))
            {
                return false;
            }

            // Check if the pattern contains typical regex metacharacters
            if (!ContainsRegexMetaCharacters(pattern))
            {
                return false;
            }

            try
            {
                // Try to create a Regex object from the pattern
                Regex.Match(string.Empty, pattern);
            }
            catch (ArgumentException)
            {
                // If an ArgumentException is thrown, the pattern is not a valid regex
                return false;
            }

            return true;
        }
        private static bool ContainsRegexMetaCharacters(string pattern)
        {
            // List of common regex metacharacters
            string[] regexMetaCharacters = { ".", "^", "$", "*", "+", "?", "(", ")", "[", "]", "{", "}", "\\", "|" };

            foreach (var metaChar in regexMetaCharacters)
            {
                if (pattern.Contains(metaChar))
                {
                    return true;
                }
            }

            return false;
        }

        private string ExtractImageUrl(XElement parent, List<string> imageTags)
        {
            //Regex for UrlToImage!!
            //var checkRegex = new Regex(imageTags[1]);
            //var check = IsValidRegex(imageTags[1]);
            if (imageTags[1] != "")
            {
                if (!IsValidRegex(imageTags[1]))
                {
                    // Scenario 2: Single XML tag with attribute
                    var xmlTag = imageTags[0].Trim();

                    var xmlTagElement = FindElementByTagName(parent, xmlTag);

                    if (xmlTagElement != null)
                    {
                        var xmlAttribute = imageTags[1].Trim(); // Get the attribute name dynamically
                        var attributeValue = xmlTagElement.Attribute(xmlAttribute)?.Value ?? xmlTagElement.Value;

                        if (!string.IsNullOrEmpty(attributeValue))
                        {
                            return attributeValue;
                        }
                    }
                }
                else // if (imageTags.Count == 2 && !string.IsNullOrEmpty(imageTags[1].Trim()))
                {
                    // Scenario 3: Tag with regex
                    var xmlTag = imageTags[0].Trim();
                    var regexPattern = imageTags[1].Trim();

                    var xmlTagElement = FindElementByTagName(parent, xmlTag);

                    if (xmlTagElement != null)
                    {
                        var attributeValue = xmlTagElement.Attribute("url")?.Value ?? xmlTagElement.Value;

                        if (!string.IsNullOrEmpty(attributeValue))
                        {
                            var regex = new Regex(regexPattern, RegexOptions.IgnoreCase);
                            var match = regex.Match(attributeValue);

                            if (match.Success && match.Groups.Count > 1)
                            {
                                return match.Groups[1].Value;
                            }
                        }
                    }
                }
            }
            else // if (imageTags.Count == 1)
            {
                // Scenario 1: One XML tag
                var xmlTag = imageTags[0].Trim();
                //var xmlTag2 = imageTags[1].Trim();

                var xmlTagElement = FindElementByTagName(parent, xmlTag);
                //var xmlTagElement2 = FindElementByTagName(parent, xmlTag2);

                if (xmlTagElement != null)// && xmlTagElement2 != null)
                {
                    return xmlTagElement?.Value;// ?? xmlTagElement2.Value;
                }
            }

            return string.Empty;
        }

        private string ExtractImageUrl1(XElement parent, List<string> imageTags)
        {
            var regexPattern = imageTags[1].Trim();
            // Scenario 1: Two XML tags
            if (imageTags[1] == null)
            {
                var xmlTag = imageTags[0].Trim();
                var xmlTagElement = FindElementByTagName(parent, xmlTag);
            
                if (xmlTagElement != null)
                {
                    var attributeValue = xmlTagElement.Attribute("url")?.Value ?? xmlTagElement.Value;
                    return attributeValue;
                }
            }
            // Scenario 3: Tag with regex
            else if (!string.IsNullOrEmpty(regexPattern))
            {
                //var checkRegex = new Regex(regexPattern);
                var match1 = Regex.Match(imageTags[1], imageTags[1], RegexOptions.IgnoreCase);
                
                if (!match1.Success)
                {
                    // does not match
                    // Scenario 2: Single XML tag with attribute
                    var xmlTag = imageTags[0].Trim();
                    var xmlAttribute = imageTags[1].Trim();

                    var xmlTagElement = FindElementByTagName(parent, xmlTag);
                    //var xmlTagElement2 = FindElementByTagName(parent, xmlTag2);

                    if (xmlTagElement != null)// && xmlTagElement2 != null)
                    {
                        return xmlTagElement.Attribute(xmlAttribute)?.Value; // ?? xmlTagElement2.Value;
                    }
                }
                else
                {
                    // does match
                    // Scenario 3: Tag with regex
                    var xmlTag = imageTags[0].Trim();
                    //var regexPattern = imageTags[1].Trim();

                    var xmlTagElement = FindElementByTagName(parent, xmlTag);
                    if (xmlTagElement != null)
                    {
                        var attributeValue = xmlTagElement.Attribute("url")?.Value ?? xmlTagElement.Value;
                        if (!string.IsNullOrEmpty(attributeValue))
                        {
                            var regex = new Regex(regexPattern, RegexOptions.IgnoreCase);
                            var match = regex.Match(attributeValue);
                            if (match.Success && match.Groups.Count > 1)
                            {
                                return match.Groups[1].Value;
                            }
                        }
                    }
                }
                
            }
            // Scenario 2: Single XML tag with attribute
            else
            {
                //var xmlTag = imageTags[0].Trim();
                //var xmlAttribute = imageTags[1].Trim();

                //var xmlTagElement = FindElementByTagName(parent, xmlTag);
                ////var xmlTagElement2 = FindElementByTagName(parent, xmlTag2);

                //if (xmlTagElement != null)// && xmlTagElement2 != null)
                //{
                //    return xmlTagElement.Attribute(xmlAttribute)?.Value; // ?? xmlTagElement2.Value;
                //}
            }
            //foreach (var tag in imageTags)
            //{
            //    var parts = tag.Split(new [] { ',' }, 2); // Split into max 2 parts

            //    if (parts.Length == 2)
            //    {
            //        // Scenario 3: Tag with regex
            //        var xmlTag = parts[0].Trim();
            //        var regexPattern = parts[1].Trim();

            //        var xmlTagElement = FindElementByTagName(parent, xmlTag);
            //        if (xmlTagElement != null)
            //        {
            //            var attributeValue = xmlTagElement.Attribute("url")?.Value ?? xmlTagElement.Value;
            //            if (!string.IsNullOrEmpty(attributeValue))
            //            {
            //                var regex = new Regex(regexPattern, RegexOptions.IgnoreCase);
            //                var match = regex.Match(attributeValue);
            //                if (match.Success && match.Groups.Count > 1)
            //                {
            //                    return match.Groups[1].Value;
            //                }
            //            }
            //        }
            //    }
            //    else if (parts.Length == 1)
            //    {
            //        var xmlTag = parts[0].Trim();
            //        var xmlTagElement = FindElementByTagName(parent, xmlTag);
            
            //        if (xmlTagElement != null)
            //        {
            //            var attributeValue = xmlTagElement.Attribute("url")?.Value ?? xmlTagElement.Value;
            //            return attributeValue;
            //        }
            //    }
            //    else if (parts.Length == 2 && string.IsNullOrEmpty(parts[1].Trim()))
            //    {
            //        var xmlTag1 = parts[0].Trim();
            //        var xmlTag2 = parts[1].Trim();

            //        var xmlTagElement1 = FindElementByTagName(parent, xmlTag1);
            //        var xmlTagElement2 = FindElementByTagName(parent, xmlTag2);

            //        if (xmlTagElement1 != null && xmlTagElement2 != null)
            //        {
            //            return xmlTagElement1.Attribute(xmlTag2)?.Value ?? xmlTagElement2.Value;
            //        }
            //    }
            //}

            return string.Empty;
        }

        private string StripHtmlTags(string text, List<string> allowedTags = null)
        {
            if (allowedTags == null || allowedTags.Count == 0)
            {
                return Regex.Replace(text, "<.*?>", string.Empty);
            }

            var tagList = string.Join("|", allowedTags);
            var regex = new Regex($"<(?!/?(?:{tagList})\\b)[^>]*>", RegexOptions.IgnoreCase);
            return regex.Replace(text, string.Empty);
        }

        private string RemoveScriptTags(string xmlData)
        {
            return Regex.Replace(xmlData, "<script.*?</script>", "", RegexOptions.IgnoreCase | RegexOptions.Singleline);
        }
    }
}
